File: strategy/lua/marvin/test/move/chiptc.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local ChipTC = Class("Test.Move.ChipTC", require "group/move/base")

local World = require "../base/world"
local G = World.Geometry
local MoveToPos = require "task/shared/movetopos"
local Pass = require "task/shared/pass"

ChipTC.MIN_ROBOTS = 1
ChipTC.MAX_ROBOTS = 1

local pos_y = -G.FieldHeightHalf + 0.1
local positions = {
	Vector(-G.FieldWidthHalf + 0.1, pos_y),
	Vector(-G.FieldWidthHalf / 2 + 0.05, pos_y),
	Vector(0, pos_y)
}

local distances = {
	1.5,
	2.0,
	2.5,
	3.0,
	3.5
}

-- ==========================================

local CURRENT_POS = positions[1]

-- ==========================================


function ChipTC.canStart()
	return true
end

function ChipTC:_init()
	self._distance = nil
	self._recalculate = true
end

function ChipTC:_canContinue()
	return true
end

function ChipTC:_updateTasks()
	local taskAssignments = {}

	if World.RefereeState == "DirectOffensive" then
		if self._recalculate then
			self._distance = distances[math.random(5)]
			self._recalculate = false
			log(self._distance)
		end
	else
		self._recalculate = true
	end

	if World.RefereeState == "DirectOffensive" then
		local ballPos = CURRENT_POS + Vector(0, self._robots[1].shootRadius + World.Ball.radius)
		local target = ballPos + Vector(0, self._distance * 2.5)
		taskAssignments[self._robots[1]] = { class = Pass, params = {self._robots[1], target, true, self._distance * 2.5} }
	else
		taskAssignments[self._robots[1]] = { class = MoveToPos, params = { CURRENT_POS, math.pi/2 } }
	end

	return taskAssignments, self._robots[1]
end

return ChipTC

File: strategy/typescript/glados/test/move/chiptc.ts
let ChipTC = Class("Test.Move.ChipTC", require "group/move/base")


import * as MathUtil from "base/mathutil";
import * as World from "base/world";
let G = World.Geometry
import {MoveToPos} from "glados/task/shared/movetopos";
import {Pass} from "glados/task/shared/pass";

ChipTC.MIN_ROBOTS = 1
ChipTC.MAX_ROBOTS = 1

let pos_y = -G.FieldHeightHalf + 0.1
let positions = [
	new Vector(-G.FieldWidthHalf + 0.1, pos_y),
	new Vector(-G.FieldWidthHalf / 2 + 0.05, pos_y),
	new Vector(0, pos_y)
];

let distances = {
	1.5,
	2.0,
	2.5,
	3.0,
	3.5
}

// ==========================================

let CURRENT_POS = positions[1]

// ==========================================


function ChipTC.canStart () {
	return true;
}

function ChipTC:_init () {
	this._distance = nil;
	this._recalculate = true;
}

function ChipTC:_canContinue () {
	return true;
}

function ChipTC:_updateTasks () {
	let taskAssignments = {};

	if (World.RefereeState == "DirectOffensive") {
		if (this._recalculate) {
			this._distance = distances[MathUtil.randomInt([1,5])];
			this._recalculate = false;
			log(this._distance);
		}
	} else {
		this._recalculate = true;
	}

	if (World.RefereeState == "DirectOffensive") {
		let ballPos = CURRENT_POS + new Vector(0, this._robots[0].shootRadius + World.Ball.radius)
		let target = ballPos + new Vector(0, this._distance * 2.5)
		taskAssignments[this._robots[0]] = { class: Pass, params: {this._robots[0], target, true, this._distance * 2.5} }
	} else {
		taskAssignments[this._robots[0]] = { class: MoveToPos, params: { CURRENT_POS, Math.PI/2 } }
	}

	return taskAssignments, this._robots[0]
}

return ChipTC
